File: WpfControls/Directives/Directives.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;

namespace WpfFs.Directives{
    public class Ng : Control{
        public static void SetBlockMark(DependencyObject el, bool value){
            var ui = el as UIElement;
            if (ui == null)
                return;
            if (value)
                ui.MouseRightButtonDown += onMouseRightButtonDown;
            else
                ui.MouseRightButtonDown -= onMouseRightButtonDown;
        }
        static void onMouseRightButtonDown(object sender, MouseButtonEventArgs e){
            Console.WriteLine("Source = {0}, OriginalSource = {1} @ {2}", e.Source.GetType().Name, e.OriginalSource.GetType().Name, e.Timestamp);

            var source = e.OriginalSource as Control;
            if (source == null)
                return;
            var showThickness = new Thickness(5);
            if (source.BorderThickness != showThickness){
                source.BorderThickness = showThickness;
                source.BorderBrush = Brushes.Black;
            }else
                source.BorderThickness = new Thickness(0);
        }
    }
}

File: WpfFs/Directives.fs
namespace WpfFs.Directives

open System
open System.Windows

type Ng() =
    static let onMouseRightButtonDown (sender: obj) (e: Input.MouseButtonEventArgs) =
        printfn "Source = %A, OriginalSource = %A @ %A" (e.Source.GetType().Name) (e.OriginalSource.GetType().Name) (e.Timestamp)

        match e.Source with
        | :? Controls.Control as source ->
                if source.BorderThickness <> Thickness(5.0) then
                    source.BorderThickness <- Thickness(5.0)
                    source.BorderBrush <- Media.Brushes.Black
                else
                    source.BorderThickness <- Thickness(0.0)
        | _ -> ()

    static member SetBlockMark(el: DependencyObject, value: bool) =
        match el with
        | :? UIElement as ui -> if value then
                                    ui.AddHandler(UIElement.MouseRightButtonDownEvent, Input.MouseButtonEventHandler(onMouseRightButtonDown), true)
                                else
                                    ui.RemoveHandler(UIElement.MouseRightButtonDownEvent, Input.MouseButtonEventHandler(onMouseRightButtonDown))
        | _ -> ()